File: System/Protocol/Allors.Protocol.Json/Data/Procedure.cs
// <copyright file="Pull.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Protocol.Json.Data
{
    using System.Collections.Generic;
    using System.Text.Json.Serialization;

    public class Procedure : IVisitable
    {
        [JsonPropertyName("name")]
        public string Name { get; set; }

        [JsonPropertyName("namedCollections")]
        public string[][] CollectionByName { get; set; }

        [JsonPropertyName("namedObjects")]
        public string[][] ObjectByName { get; set; }

        [JsonPropertyName("namedValues")]
        public string[][] ValueByName { get; set; }

        [JsonPropertyName("objectVersions")]
        public string[][] VersionByObject { get; set; }

        public void Accept(IVisitor visitor) => visitor.VisitProcedure(this);
    }
}

File: Core/Workspace/Typescript/libs/data/core/src/lib/Procedure.ts
export type ProcedureArgs = Pick<Procedure, 'name' | 'namedCollections' | 'namedObjects' | 'namedValues' | 'objectVersions'>;

export class Procedure {
  name: string;

  namedCollections?: string[][];

  namedObjects?: string[][];

  namedValues?: string[][];

  objectVersions?: string[][];

  constructor(args: ProcedureArgs) {
    Object.assign(this, args);
  }

  public toJSON(): any {
    return {
      name: this.name,
      namedCollections: this.namedCollections,
      namedObjects: this.namedObjects,
      namedValues: this.namedValues,
      objectVersions: this.objectVersions,
    };
  }
}
